lib/stitches/auto_fill.py

```diff
@@ -128,248 +131,241 @@ def build_graph(shape, segments, angle, row_spacing, max_stitch_length):
+    """
 
-def nearest_node_on_outline(graph, point, outline_index=0):
-    point = shapely.geometry.Point(*point)
-    outline_nodes = [node for node, data in graph.nodes(data=True) if data['index'] == outline_index]
-    nearest = min(outline_nodes, key=lambda node: shapely.geometry.Point(*node).distance(point))
+    # If the shape is smaller, we'll have less room to maneuver and it's more likely
+    # we'll travel around the outside border of the shape.  Counteract that by making
+    # the grid denser.
+    if shape.area < 10000:
+        scale = 0.5
+    else:
+        scale = 1.0
 
-    return nearest
+    grating1 = shgeo.MultiLineString(
+        list(chain(*intersect_region_with_grating(shape, fill_angle + math.pi / 4, scale * 2 * PIXELS_PER_MM))))
+    grating2 = shgeo.MultiLineString(
+        list(chain(*intersect_region_with_grating(shape, fill_angle - math.pi / 4, scale * 2 * PIXELS_PER_MM))))
+    grating3 = shgeo.MultiLineString(
+        list(chain(*intersect_region_with_grating(shape, fill_angle - math.pi / 2, scale * math.sqrt(2) * PIXELS_PER_MM))))
 
+    debug.add_layer("auto-fill travel")
+    debug.log_line_strings(grating1, "grating1")
+    debug.log_line_strings(grating2, "grating2")
+    debug.log_line_strings(grating3, "grating3")
 
-def get_outline_nodes(graph, outline_index=0):
-    outline_nodes = [(node, data['projection'])
-                     for node, data
-                     in graph.nodes(data=True)
-                     if data['index'] == outline_index]
-    outline_nodes.sort(key=lambda node_projection: node_projection[1])
-    outline_nodes = [node for node, data in outline_nodes]
+    endpoints = [coord for mls in (grating1, grating2, grating3)
+                 for ls in mls
+                 for coord in ls.coords]
 
-    return outline_nodes
+    diagonal_edges = grating1.symmetric_difference(grating2)
 
+    # without this, floating point inaccuracies prevent the intersection points from lining up perfectly.
+    vertical_edges = snap(grating3.difference(grating1), diagonal_edges, 0.005)
 
-def find_initial_path(graph, starting_point, ending_point=None):
-    starting_node = nearest_node_on_outline(graph, starting_point)
+    return endpoints, chain(diagonal_edges, vertical_edges)
 
-    if ending_point is not None:
-        ending_node = nearest_node_on_outline(graph, ending_point)
 
-    if ending_point is None or starting_node is ending_node:
-        # If they didn't give an ending point, pick either neighboring node
-        # along the outline -- doesn't matter which.  We do this because
-        # the algorithm requires we start with _some_ path.
-        neighbors = [n for n, keys in graph.adj[starting_node].iteritems() if 'outline' in keys]
-        return [PathEdge((starting_node, neighbors[0]), "initial")]
-    else:
-        outline_nodes = get_outline_nodes(graph)
+def check_graph(graph, shape, max_stitch_length):
+    if networkx.is_empty(graph) or not networkx.is_eulerian(graph):
```

Comment:
I would write:

```
error_msg = "This shape is so small that it cannot be filled with rows of stitches.  "
                     "It would probably look best as a satin column or running stitch."
                      if shape.area < max_stitch_length ** 2 else
                      "Cannot parse shape. This most often happens because your shape is "
                      "made up of multiple sections that aren't connected."
raise InvalidPath(_(error_msg))
```

Reply:
Done, thanks.